diff --git a/src/linebreak.c b/src/linebreak.c
index 0c6514f..b1ecf3d 100644
--- a/src/linebreak.c
+++ b/src/linebreak.c
@@ -744,6 +744,12 @@ int lb_process_next_char(
 
     return brk;
 }
+enum LineBreakClass lb_get_char_class(
+        struct LineBreakContext *lbpCtx,
+        utf32_t ch)
+{
+    return get_char_lb_class_lang(ch, lbpCtx->lbpLang);
+}
 
 /**
  * Sets the line breaking information for a generic input string.
diff --git a/src/linebreakdef.h b/src/linebreakdef.h
index 8ef012a..03c1805 100644
--- a/src/linebreakdef.h
+++ b/src/linebreakdef.h
@@ -179,5 +179,8 @@ size_t set_linebreaks(
         enum BreakOutputType outputType,
         char *brks,
         get_next_char_t get_next_char);
+enum LineBreakClass lb_get_char_class(
+        struct LineBreakContext *lbpCtx,
+        utf32_t ch);
 
 #endif /* LINEBREAKDEF_H */
