diff --git i/lib/hashlib/CMakeLists.txt w/lib/hashlib/CMakeLists.txt
index 63d3852..2f7880c 100644
--- i/lib/hashlib/CMakeLists.txt
+++ w/lib/hashlib/CMakeLists.txt
@@ -11,7 +11,6 @@ include_directories(${PROJECT_SOURCE_DIR}/include)
 
 set(HASHLIB_SRCS
     ${PROJECT_SOURCE_DIR}/src/crc32.cpp
-    ${PROJECT_SOURCE_DIR}/src/digest.cpp
     ${PROJECT_SOURCE_DIR}/src/keccak.cpp
     ${PROJECT_SOURCE_DIR}/src/md5.cpp
     ${PROJECT_SOURCE_DIR}/src/sha1.cpp
