diff --git a/CMakeLists.txt b/CMakeLists.txt
index 68b5343..fbc2e01 100644
--- a/lib/cpr/CMakeLists.txt
+++ b/lib/cpr/CMakeLists.txt
@@ -132,7 +132,7 @@ if(CPR_ENABLE_SSL)
             set(CPR_BUILD_TESTS_SSL OFF)
             set(SSL_BACKEND_USED "DarwinSSL")
         else()
-            find_package(OpenSSL)
+            find_package(OpenSSL COMPONENTS Crypto SSL)
             if(OPENSSL_FOUND)
                 message(STATUS "SSL auto detect: Using OpenSSL.")
                 set(SSL_BACKEND_USED "OpenSSL")
@@ -147,7 +147,7 @@ if(CPR_ENABLE_SSL)
         endif()
     else()
         if(CPR_FORCE_OPENSSL_BACKEND)
-            find_package(OpenSSL)
+            find_package(OpenSSL COMPONENTS Crypto SSL)
             if(OPENSSL_FOUND)
                 message(STATUS "Using OpenSSL.")
                 set(SSL_BACKEND_USED "OpenSSL")
@@ -171,7 +171,7 @@ endif()
 
 if(SSL_BACKEND_USED STREQUAL "OpenSSL")
 # Fix missing OpenSSL includes for Windows since in 'ssl_ctx.cpp' we include OpenSSL directly
-find_package(OpenSSL REQUIRED)
+find_package(OpenSSL REQUIRED COMPONENTS Crypto SSL)
     add_compile_definitions(OPENSSL_BACKEND_USED)
 endif()
 
