--- a/meson.build
+++ b/meson.build
@@ -320,11 +320,16 @@ else
 endif
 
 # BZip2 support.
-bzip2_dep = dependency('bzip2', required: false)
+bzip2_dep = dependency(
+  'bzip2',
+  required: get_option('bzip2').disabled() ? get_option('bzip2') : false,
+)
 if not bzip2_dep.found()
-  bzip2_dep = cc.find_library('bz2',
+  bzip2_dep = cc.find_library(
+    'bz2',
     has_headers: ['bzlib.h'],
-    required: get_option('bzip2'))
+    required: get_option('bzip2'),
+  )
 endif
 
 if bzip2_dep.found()
