diff --git a/libarchive/archive_entry.h b/libarchive/archive_entry.h
index aa4f4dfb..17cdd2e5 100644
--- a/libarchive/archive_entry.h
+++ b/libarchive/archive_entry.h
@@ -93,7 +93,7 @@ typedef ssize_t la_ssize_t;
 #endif
 
 /* Get a suitable definition for mode_t */
-#if ARCHIVE_VERSION_NUMBER >= 3999000
+#if 1 //ARCHIVE_VERSION_NUMBER >= 3999000
 /* Switch to plain 'int' for libarchive 4.0.  It's less broken than 'mode_t' */
 # define	__LA_MODE_T	int
 #elif defined(_WIN32) && !defined(__CYGWIN__) && !defined(__BORLANDC__) && !defined(__WATCOMC__)
