--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -8,7 +8,7 @@
 #
 # ##############################################################################
 
-cmake_minimum_required(VERSION 3.10 FATAL_ERROR)
+cmake_minimum_required(VERSION 3.15 FATAL_ERROR)
 
 # In-source builds are disabled.
 if("${CMAKE_CURRENT_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_BINARY_DIR}")
@@ -407,6 +407,7 @@
   endif(NOT Leptonica_FOUND)
   include_directories(${Leptonica_INCLUDE_DIRS})
 
+  if(NOT DISABLE_TIFF)
   check_leptonica_tiff_support()
   if ((NOT LEPT_TIFF_RESULT EQUAL 0) AND LEPT_TIFF_COMPILE_SUCCESS)
     message(NOTICE "Leptonica was build without TIFF support! Disabling TIFF support...")
@@ -414,6 +415,7 @@
   elseif(NOT ${CMAKE_VERSION} VERSION_LESS "3.25")
     message(STATUS "Leptonica was build with TIFF support.")
   endif()
+  endif()
 
   # Check for optional libraries.
   if(DISABLE_TIFF)
@@ -827,6 +829,11 @@
     PRIVATE -DTESS_EXPORTS
     INTERFACE -DTESS_IMPORTS)
   # generate_export_header          (libtesseract EXPORT_MACRO_NAME TESS_API)
+else()
+    # Enable `-fvisibility=hidden`.
+    set_target_properties(libtesseract PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    # Enable `-fvisibility-inlines-hidden`.
+    set_target_properties(libtesseract PROPERTIES VISIBILITY_INLINES_HIDDEN TRUE)
 endif()
 target_link_libraries(libtesseract PRIVATE ${LIB_Ws2_32} ${LIB_pthread})
 if(OpenMP_CXX_FOUND)
@@ -851,7 +858,7 @@
                           ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH})
 set_target_properties(
   libtesseract PROPERTIES SOVERSION
-                          ${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH})
+                          ${VERSION_MAJOR})
 
 set_target_properties(
   libtesseract
@@ -887,12 +894,13 @@
 endif()
 
 if(ANDROID)
-  add_definitions(-DANDROID)
+  add_definitions(-DANDROID -DHAVE_ANDROID_GETCPUFAMILY)
   find_package(CpuFeaturesNdkCompat REQUIRED)
   target_include_directories(
     libtesseract
     PRIVATE "${CpuFeaturesNdkCompat_DIR}/../../../include/ndk_compat")
   target_link_libraries(libtesseract PRIVATE CpuFeatures::ndk_compat)
+  set(cpu_features_LIBS "-lcpu_features -lndk_compat")
 endif()
 
 # ##############################################################################
--- i/tesseract.pc.cmake
+++ w/tesseract.pc.cmake
@@ -8,6 +8,6 @@
 URL: https://github.com/tesseract-ocr/tesseract
 Version: @tesseract_VERSION@
 Requires.private: lept
-Libs: -L${libdir} -l@tesseract_OUTPUT_NAME@ @libarchive_LIBS@ @libcurl_LIBS@ @TENSORFLOW_LIBS@
+Libs: -L${libdir} -l@tesseract_OUTPUT_NAME@ @cpu_features_LIBS@ @libarchive_LIBS@ @libcurl_LIBS@ @TENSORFLOW_LIBS@
 Libs.private:
 Cflags: -I${includedir}

