--- a/tesseract_mod/tesscapi.cpp
+++ b/tesseract_mod/tesscapi.cpp
@@ -1,4 +1,3 @@
-#include "config_auto.h"
 /*
 ** tesscapi.cpp    willus.com attempt at C wrapper for tesseract.
 **                 (Butchered from tesseractmain.cpp)
@@ -21,7 +20,6 @@
 **
 */
 
-#include "config_auto.h"
 /*
 #include "mfcpch.h"
 */
@@ -44,11 +42,6 @@
 
 #include "allheaders.h"
 #include "baseapi.h"
-#include "strngs.h"
-#include "params.h"
-#include "blobs.h"
-// For our try/catch block
-#include <stdexcept>
 /*
 #include "notdll.h"
 */
@@ -56,6 +49,9 @@
 /* C Wrappers */
 #include "tesseract.h"
 
+// For our try/catch block
+#include <stdexcept>
+
 // static tesseract::TessBaseAPI api[4];
 
 /*
