diff --git a/aes/Makefile b/aes/Makefile
new file mode 100644
index 0000000..76ab5f1
--- /dev/null
+++ b/aes/Makefile
@@ -0,0 +1,20 @@
+CC?=cc
+CFLAGS?=-O
+EXTRA_CFLAGS:=-DHAVE_AES
+OBJS=aescrypt.o aeskey.o aestab.o fileenc.c hmac.o prng.c pwd2key.o sha1.o
+ARFLAGS=rv
+
+.c.o:
+	$(CC) -c $(CFLAGS) $(EXTRA_CFLAGS) $*.c
+
+libaes.a: $(OBJS)
+	$(ECHO) $(AR) $(ARFLAGS) ../libaes.a $?
+	$(AR) $(ARFLAGS) ../libaes.a $?
+	$(RANLIB) ../libaes.a
+
+all: libaes.a
+
+.PHONY: clean
+
+clean:
+	rm -f *.o *.a
