diff --git a/Makefile b/Makefile
index 68cabaf..e43dea0 100644
--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,7 @@ COMPAT_DIR=lua-compat-5.2
 SRCS=$(wildcard $(SRC_DIR)/*.c)
 OBJS=$(SRCS:%.c=%.o)
 
-CFLAGS=-I$(LUA_DIR)/src -I$(COMPAT_DIR)/c-api -shared -fPIC
+LUA_CFLAGS:= -I$(LUA_DIR)/src -I$(COMPAT_DIR)/c-api -shared -fPIC
 
 ifndef DARWIN
 	CFLAGS+= -Wl,-E -Wl,-rpath,'$$ORIGIN'
@@ -38,8 +38,8 @@ LIBSERIALIZE=$(OUTPUT_DIR)/serialize.$(if $(WIN32),dll,so)
 libs: $(LIBLUACOMPAT52) $(LIBSERIALIZE)
 
 $(LIBLUACOMPAT52): $(COMPAT_DIR)/c-api/compat-5.2.c $(LUA_DIR)
-	$(CC) $(CFLAGS) -o $@ $< -L$(OUTPUT_DIR) $(LDFLAGS)
+	$(CC) $(CFLAGS) $(LUA_CFLAGS) -o $@ $< $(LDFLAGS) -L$(OUTPUT_DIR)
 
 $(LIBSERIALIZE): $(SRC_DIR)/serialize.c $(LUA_DIR)
-	$(CC) $(CFLAGS) -o $@ $< -lluacompat52 -L$(OUTPUT_DIR) $(LDFLAGS)
+	$(CC) $(CFLAGS) $(LUA_CFLAGS) -o $@ $< $(LDFLAGS) -lluacompat52 -L$(OUTPUT_DIR)
 
