--- i/Makefile
+++ w/Makefile
@@ -14,6 +14,9 @@ else
 	STRIP?=strip
 endif
 
+# For use in recursive calls we don't want `make -n` to follow through.
+SUBMAKE = $(MAKE)
+
 DEBUG_CFLAGS:=-Og -fno-omit-frame-pointer -pipe -g
 # Fallback CFLAGS, we honor the env first and foremost!
 OPT_CFLAGS:=-O2 -fomit-frame-pointer -pipe
@@ -197,7 +200,7 @@ libevdev.built:
 	--prefix="$(CURDIR)/libevdev-staged" \
 	--enable-static \
 	--disable-shared && \
-	$(MAKE) install
+	$(SUBMAKE) install
 	touch libevdev.built
 
 fbink.built:
