--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -1,4 +1,4 @@
-cmake_minimum_required (VERSION 3.16.4)
+cmake_minimum_required (VERSION 3.17.5)
 if(MSVC)
 	cmake_policy(SET CMP0091 NEW)
 endif()
@@ -532,8 +532,8 @@
 		# Create pkgconfig files.
 		set(prefix      ${CMAKE_INSTALL_PREFIX})
 		set(exec_prefix \${prefix})
-		set(libdir      \${exec_prefix}/${CMAKE_INSTALL_LIBDIR})
-		set(includedir  \${prefix}/${CMAKE_INSTALL_INCLUDEDIR})
+		set(libdir      \${exec_prefix}/lib)
+		set(includedir  \${prefix}/include)
 		if(PLATFORM_LIBS)
 			string(REGEX REPLACE ";" " -l" PLATFORM_LDADD ";${PLATFORM_LIBS}")
 		endif()
