--- i/deps/turbo_ffi_wrap.c
+++ w/deps/turbo_ffi_wrap.c
@@ -58,7 +58,7 @@
     int common_name_loc = -1;
     X509_NAME_ENTRY *common_name_entry = 0;
     ASN1_STRING *common_name_asn1 = 0;
-    char *common_name_str = 0;
+    const char *common_name_str = 0;
 
     common_name_loc = X509_NAME_get_index_by_NID(X509_get_subject_name(
                                                      (X509 *) server_cert),
@@ -77,7 +77,7 @@
     if (!common_name_asn1) {
         return Error;
     }
-    common_name_str = (char *) ASN1_STRING_data(common_name_asn1);
+    common_name_str = (const char *) ASN1_STRING_get0_data(common_name_asn1);
     if (ASN1_STRING_length(common_name_asn1) != strlen(common_name_str)) {
         return MalformedCertificate;
     }
@@ -113,7 +113,7 @@
     for (i=0; i<san_names_nb; i++){
         const GENERAL_NAME *current_name = sk_GENERAL_NAME_value(san_names, i);
         if (current_name->type == GEN_DNS){
-            char *dns_name = (char *)ASN1_STRING_data(current_name->d.dNSName);
+            const char *dns_name = (const char *)ASN1_STRING_get0_data(current_name->d.dNSName);
             dns_name_sz = strlen(dns_name);
             if (ASN1_STRING_length(current_name->d.dNSName) != dns_name_sz){
                 result = MalformedCertificate;
