diff --git a/src/hb.hh b/src/hb.hh
index 47b710e67..5ddaa031f 100644
--- a/src/hb.hh
+++ b/src/hb.hh
@@ -481,6 +481,10 @@ static int HB_UNUSED _hb_errno = 0;
 #define HB_NO_SETLOCALE 1
 #endif
 
+#if !defined(HB_NO_SETLOCALE) && defined(__ANDROID__) && __ANDROID_API__ < 21
+#define HB_NO_SETLOCALE 1
+#endif
+
 #ifndef HB_NO_SETLOCALE
 
 #include <locale.h>
diff --git a/src/hb-buffer-verify.cc b/src/hb-buffer-verify.cc
index 76f821637..57a619c6b 100644
--- a/src/hb-buffer-verify.cc
+++ b/src/hb-buffer-verify.cc
@@ -46,11 +46,13 @@ buffer_verify_error (hb_buffer_t *buffer,
 {
   va_list ap;
   va_start (ap, fmt);
+#ifndef HB_NO_BUFFER_MESSAGE
   if (buffer->messaging ())
   {
     buffer->message_impl (font, fmt, ap);
   }
   else
+#endif
   {
     fprintf (stderr, "harfbuzz ");
     vfprintf (stderr, fmt, ap);
