diff --git a/src/hb-atomic.hh b/src/hb-atomic.hh
index f9afd4ff..783a4198 100644
--- a/src/hb-atomic.hh
+++ b/src/hb-atomic.hh
@@ -49,7 +49,7 @@
 /* Defined externally, i.e. in config.h. */
 
 
-#elif !defined(HB_NO_MT) && defined(__ATOMIC_ACQUIRE)
+#elif !defined(HB_NO_MT) && defined(__ATOMIC_ACQUIRE) && defined(__NOPEY_NOPE)
 
 /* C++11-style GCC primitives. */
 
@@ -72,7 +72,7 @@ _hb_atomic_ptr_impl_cmplexch (const void **P, const void *O_, const void *N)
 }
 #define hb_atomic_ptr_impl_cmpexch(P,O,N)	_hb_atomic_ptr_impl_cmplexch ((const void **) (P), (O), (N))
 
-#elif !defined(HB_NO_MT) && __cplusplus >= 201103L
+#elif !defined(HB_NO_MT) && __cplusplus >= 201103L && defined(__NOPEY_NOPE)
 
 /* C++11 atomics. */
 
