diff --git a/src/video/cocoa/SDL_cocoaevents.m b/src/video/cocoa/SDL_cocoaevents.m
index 39c5c4f38..b27306629 100644
--- a/src/video/cocoa/SDL_cocoaevents.m
+++ b/src/video/cocoa/SDL_cocoaevents.m
@@ -55,6 +55,18 @@ static SDL_Window *FindSDLWindowForNSWindow(NSWindow *win)
     return sdlwindow;
 }
 
+static void _openUrl(NSString *url) {
+    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:url]];
+}
+
+static void _openFile(NSString *file, NSString *app) {
+    NSString *path = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:file];
+    if (path) {
+        if (!app) app = @"TextEdit";
+        [[NSWorkspace sharedWorkspace] openFile:path withApplication:app];
+    }
+}
+
 @interface SDLApplication : NSApplication
 
 - (void)terminate:(id)sender;
@@ -66,6 +78,23 @@ static SDL_Window *FindSDLWindowForNSWindow(NSWindow *win)
 
 @implementation SDLApplication
 
+/* KOReader specific actions */
+- (IBAction) openWeb: (NSMenuItem*) sender {
+    _openUrl(@"https://koreader.rocks");
+}
+
+- (IBAction) openWiki: (NSMenuItem*) sender {
+    _openUrl(@"https://github.com/koreader/koreader/wiki");
+}
+
+- (IBAction) openForum: (NSMenuItem*) sender {
+    _openUrl(@"https://www.mobileread.com/forums/forumdisplay.php?f=276");
+}
+
+- (IBAction) openLicense: (NSMenuItem*) sender {
+    _openFile(@"COPYING", nil);
+}
+
 // Override terminate to handle Quit and System Shutdown smoothly.
 - (void)terminate:(id)sender
 {
@@ -284,30 +313,7 @@ static void Cocoa_DispatchEvent(NSEvent *theEvent)
 
 - (void)applicationDidFinishLaunching:(NSNotification *)notification
 {
-    /* The menu bar of SDL apps which don't have the typical .app bundle
-     * structure fails to work the first time a window is created (until it's
-     * de-focused and re-focused), if this call is in Cocoa_RegisterApp instead
-     * of here. https://github.com/libsdl-org/SDL/issues/1913
-     */
-
-    /* this apparently became unnecessary on macOS 14.0, and will addition pop up a
-       hidden dock if you're moving the mouse during launch, so change the default
-       behaviour there.  https://github.com/libsdl-org/SDL/issues/10340
-       (13.6 still needs it, presumably 13.7 does, too.) */
-    SDL_bool background_app_default = SDL_FALSE;
-    if (@available(macOS 14.0, *)) {
-        background_app_default = SDL_TRUE;  /* by default, don't explicitly activate the dock and then us again to force to foreground */
-    }
-
-    if (!SDL_GetHintBoolean(SDL_HINT_MAC_BACKGROUND_APP, background_app_default)) {
-        /* Get more aggressive for Catalina: activate the Dock first so we definitely reset all activation state. */
-        for (NSRunningApplication *i in [NSRunningApplication runningApplicationsWithBundleIdentifier:@"com.apple.dock"]) {
-            [i activateWithOptions:NSApplicationActivateIgnoringOtherApps];
-            break;
-        }
-        SDL_Delay(300);  /* !!! FIXME: this isn't right. */
-        [NSApp activateIgnoringOtherApps:YES];
-    }
+    [NSApp activateIgnoringOtherApps:YES];
 
     /* If we call this before NSApp activation, macOS might print a complaint
      * about ApplePersistenceIgnoreState. */
@@ -500,12 +506,6 @@ void Cocoa_RegisterApp(void)
             }
         }
         [NSApp finishLaunching];
-        if ([NSApp delegate]) {
-            /* The SDL app delegate calls this in didFinishLaunching if it's
-             * attached to the NSApp, otherwise we need to call it manually.
-             */
-            [SDLApplication registerUserDefaults];
-        }
     }
     if (NSApp && !appDelegate) {
         appDelegate = [[SDLAppDelegate alloc] init];
