diff -uNr a/src/video/cocoa/SDL_cocoaevents.m e/src/video/cocoa/SDL_cocoaevents.m
--- a/src/video/cocoa/SDL_cocoaevents.m	2020-03-11 02:36:18.000000000 +0100
+++ e/src/video/cocoa/SDL_cocoaevents.m	2020-07-28 21:40:49.000000000 +0200
@@ -34,6 +34,18 @@
 #define kIOPMAssertPreventUserIdleDisplaySleep kIOPMAssertionTypePreventUserIdleDisplaySleep
 #endif
 
+static void _openUrl(NSString *url) {
+    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:url]];
+}
+
+static void _openFile(NSString *file, NSString *app) {
+    NSString *path = [[[NSBundle mainBundle] resourcePath] stringByAppendingPathComponent:file];
+    if (path) {
+        if (!app) app = @"TextEdit";
+        [[NSWorkspace sharedWorkspace] openFile:path withApplication:app];
+    }
+}
+
 @interface SDLApplication : NSApplication
 
 - (void)terminate:(id)sender;
@@ -45,6 +57,23 @@
 
 @implementation SDLApplication
 
+/* KOReader specific actions */
+- (IBAction) openWeb: (NSMenuItem*) sender {
+    _openUrl(@"https://koreader.rocks");
+}
+
+- (IBAction) openWiki: (NSMenuItem*) sender {
+    _openUrl(@"https://github.com/koreader/koreader/wiki");
+}
+
+- (IBAction) openForum: (NSMenuItem*) sender {
+    _openUrl(@"https://www.mobileread.com/forums/forumdisplay.php?f=276");
+}
+
+- (IBAction) openLicense: (NSMenuItem*) sender {
+    _openFile(@"COPYING", nil);
+}
+
 // Override terminate to handle Quit and System Shutdown smoothly.
 - (void)terminate:(id)sender
 {
@@ -233,20 +262,7 @@
 
 - (void)applicationDidFinishLaunching:(NSNotification *)notification
 {
-    /* The menu bar of SDL apps which don't have the typical .app bundle
-     * structure fails to work the first time a window is created (until it's
-     * de-focused and re-focused), if this call is in Cocoa_RegisterApp instead
-     * of here. https://bugzilla.libsdl.org/show_bug.cgi?id=3051
-     */
-    if (!SDL_GetHintBoolean(SDL_HINT_MAC_BACKGROUND_APP, SDL_FALSE)) {
-        /* Get more aggressive for Catalina: activate the Dock first so we definitely reset all activation state. */
-        for (NSRunningApplication *i in [NSRunningApplication runningApplicationsWithBundleIdentifier:@"com.apple.dock"]) {
-            [i activateWithOptions:NSApplicationActivateIgnoringOtherApps];
-            break;
-        }
-        SDL_Delay(300);  /* !!! FIXME: this isn't right. */
-        [NSApp activateIgnoringOtherApps:YES];
-    }
+    [NSApp activateIgnoringOtherApps:YES];
 
     /* If we call this before NSApp activation, macOS might print a complaint
      * about ApplePersistenceIgnoreState. */
@@ -424,12 +440,6 @@
             }
         }
         [NSApp finishLaunching];
-        if ([NSApp delegate]) {
-            /* The SDL app delegate calls this in didFinishLaunching if it's
-             * attached to the NSApp, otherwise we need to call it manually.
-             */
-            [SDLApplication registerUserDefaults];
-        }
     }
     if (NSApp && !appDelegate) {
         appDelegate = [[SDLAppDelegate alloc] init];
