--- i/exporters/pkg-config/libcrypto.pc.in
+++ w/exporters/pkg-config/libcrypto.pc.in
@@ -17,8 +17,8 @@
 Name: OpenSSL-libcrypto
 Description: OpenSSL cryptography library
 Version: {- $OpenSSL::safe::installdata::VERSION -}
-Libs: -L${libdir} -lcrypto
-Libs.private: {- join(' ', @OpenSSL::safe::installdata::LDLIBS) -}
+Libs: -L${libdir} -lcrypto {- join(' ', @OpenSSL::safe::installdata::LDLIBS) -}
+Libs.private:
 Cflags:{- $OUT = ' -I${includedir}';
           if (scalar @OpenSSL::safe::installdata::INCLUDEDIR_REL_PREFIX > 1) {
               $OUT = '';
