diff --git a/Makefile.shared b/Makefile.shared
index e8d222ac6a..c113e74d5b 100644
--- a/Makefile.shared
+++ b/Makefile.shared
@@ -151,9 +151,9 @@ DO_GNU_SO=$(CALC_VERSIONS); \
 	SHLIB_SUFFIX=; \
 	ALLSYMSFLAGS='-Wl,--whole-archive'; \
 	NOALLSYMSFLAGS='-Wl,--no-whole-archive'; \
-	SHAREDFLAGS="$(CFLAGS) $(SHARED_LDFLAGS) -shared -Wl,-Bsymbolic -Wl,-soname=$$SHLIB$$SHLIB_SOVER$$SHLIB_SUFFIX"
+	SHAREDFLAGS="$(CFLAGS) $(SHARED_LDFLAGS) -Wl,-rpath,"'$$'"ORIGIN -shared -Wl,-Bsymbolic -Wl,-soname=$$SHLIB$$SHLIB_SOVER$$SHLIB_SUFFIX"
 
-DO_GNU_APP=LDFLAGS="$(CFLAGS) -Wl,-rpath,$(LIBRPATH)"
+DO_GNU_APP=LDFLAGS="$(CFLAGS) -Wl,-rpath,"'$$'"ORIGIN/libs"
 
 #This is rather special.  It's a special target with which one can link
 #applications without bothering with any features that have anything to
