diff --git a/Configurations/shared-info.pl b/Configurations/shared-info.pl
index 47eddd6835..4fbb5d50b4 100644
--- a/Configurations/shared-info.pl
+++ b/Configurations/shared-info.pl
@@ -25,7 +25,7 @@ sub detect_gnu_cc {
 my %shared_info;
 %shared_info = (
     'gnu-shared' => {
-        shared_ldflag         => '-shared -Wl,-Bsymbolic',
+        shared_ldflag         => '-shared -Wl,-Bsymbolic -Wl,-rpath,\'$$ORIGIN\'',
         shared_sonameflag     => '-Wl,-soname=',
     },
     'linux-shared' => sub {
