diff --git a/src/Spore/Request.lua b/src/Spore/Request.lua
index eb983fb..44b4635 100644
--- a/src/Spore/Request.lua
+++ b/src/Spore/Request.lua
@@ -121,8 +121,9 @@ local function _form_data (data, debug_)
     local p = {}
     for k, v in pairs(data) do
         if sub(v, 1, 1) == '@' then
-            local fname = sub(v, 2)
-            local content = slurp(fname, debug_)
+            local fullpath = sub(v, 2)
+            local fname = fullpath:match("([^/\\]+)$")
+            local content = slurp(fullpath, debug_)
             p[#p+1] = 'content-disposition: form-data; name="' .. k .. '"; filename="' .. fname ..'"\r\n'
                    .. 'content-type: application/octet-stream\r\n\r\n'
                    .. content