diff --git a/gensignkey.c b/gensignkey.c
index 9b69545..7c212ce 100644
--- a/gensignkey.c
+++ b/gensignkey.c
@@ -139,14 +139,14 @@ int signkey_generate(enum signkey_type keytype, int bits, const char* filename,
 	if (ret == DROPBEAR_FAILURE) {
 		goto out;
 	}
-   /* Hard links are not possible and renam. */
-   if (skip_exist && access(filename, F_OK) == 0) {
-       /* Ok. */
-   } else if (rename(fn_temp, filename) < 0) {
-       dropbear_log(LOG_ERR, "Failed moving key file to %s: %s", filename,
-           strerror(errno));
-       ret = DROPBEAR_FAILURE;
 
+	/* Hack. We can't do hardlinks on vfat, so just use rename */
+	if (skip_exist && access(filename, F_OK) == 0) {
+		/* Ok. */
+	} else if (rename(fn_temp, filename) < 0) {
+		dropbear_log(LOG_ERR, "Failed moving key file to %s: %s", filename,
+		strerror(errno));
+		ret = DROPBEAR_FAILURE;
 	}
 
 out:
