From 8d0b48f16550c9bf3693b2fa683f21e8276b1b1a Mon Sep 17 00:00:00 2001
From: Matt Johnston <matt@ucc.asn.au>
Date: Sat, 3 Mar 2018 10:10:16 +0800
Subject: [PATCH] Fix CFLAGS for configure

---
 configure.ac | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/configure.ac b/configure.ac
index 6621120a..d1b26027 100644
--- a/configure.ac
+++ b/configure.ac
@@ -15,6 +15,7 @@ if test -s "`which hg`" && test -d "$srcdir/.hg"; then
 	AC_MSG_NOTICE([Source directory Mercurial base revision $hgrev])
 fi
 
+ORIGCFLAGS="$CFLAGS"
 # Checks for programs.
 AC_PROG_CC
 
@@ -35,7 +36,7 @@ AC_DEFUN(DB_TRYADDCFLAGS,
 }])
 
 # set compile flags prior to other tests
-if test -z "$OLDCFLAGS" && test "$GCC" = "yes"; then
+if test -z "$ORIGCFLAGS" && test "$GCC" = "yes"; then
 	AC_MSG_NOTICE(No \$CFLAGS set... using "-Os -W -Wall" for GCC)
 	CFLAGS="-Os -W -Wall"
 fi
