diff --git a/default_options.h b/default_options.h
index 3b75eb8..1b3588d 100644
--- a/default_options.h
+++ b/default_options.h
@@ -19,9 +19,9 @@ IMPORTANT: Some options will require "make clean" after changes */
 #define DROPBEAR_DEFADDRESS ""
 
 /* Default hostkey paths - these can be specified on the command line */
-#define DSS_PRIV_FILENAME "/etc/dropbear/dropbear_dss_host_key"
-#define RSA_PRIV_FILENAME "/etc/dropbear/dropbear_rsa_host_key"
-#define ECDSA_PRIV_FILENAME "/etc/dropbear/dropbear_ecdsa_host_key"
+#define DSS_PRIV_FILENAME "settings/SSH/dropbear_dss_host_key"
+#define RSA_PRIV_FILENAME "settings/SSH/dropbear_rsa_host_key"
+#define ECDSA_PRIV_FILENAME "settings/SSH/dropbear_ecdsa_host_key"
 
 /* Set NON_INETD_MODE if you require daemon functionality (ie Dropbear listens
  * on chosen ports and keeps accepting connections. This is the default.
@@ -44,10 +44,10 @@ IMPORTANT: Some options will require "make clean" after changes */
  * several kB in binary size however will make the symmetrical ciphers and hashes
  * slower, perhaps by 50%. Recommended for small systems that aren't doing
  * much traffic. */
-#define DROPBEAR_SMALL_CODE 1
+#define DROPBEAR_SMALL_CODE 0
 
 /* Enable X11 Forwarding - server only */
-#define DROPBEAR_X11FWD 1
+#define DROPBEAR_X11FWD 0
 
 /* Enable TCP Fowarding */
 /* 'Local' is "-L" style (client listening port forwarded via server)
@@ -239,7 +239,7 @@ Homedir is prepended unless path begins with / */
 
 /* The default file to store the daemon's process ID, for shutdown
    scripts etc. This can be overridden with the -P flag */
-#define DROPBEAR_PIDFILE "/var/run/dropbear.pid"
+#define DROPBEAR_PIDFILE "settings/SSH/dropbear.pid"
 
 /* The command to invoke for xauth when using X11 forwarding.
  * "-q" for quiet */
@@ -290,6 +290,6 @@ be overridden at runtime with -I. 0 disables idle timeouts */
 #define DEFAULT_IDLE_TIMEOUT 0
 
 /* The default path. This will often get replaced by the shell */
-#define DEFAULT_PATH "/usr/bin:/bin"
+#define DEFAULT_PATH "/sbin:/usr/sbin:/bin:/usr/bin"
 
 #endif /* DROPBEAR_DEFAULT_OPTIONS_H_ */
