diff -ur dropbear-2018.76/svr-main.c dropbear-2018.76-PATCHED/svr-main.c
--- dropbear-2018.76/svr-main.c	2018-02-27 15:25:12.000000000 +0100
+++ dropbear-2018.76-PATCHED/svr-main.c	2018-04-24 10:16:55.621898430 +0200
@@ -156,7 +156,7 @@
 			closefds = 1;
 		}
 #endif
-		if (daemon(0, closefds) < 0) {
+		if (daemon(1, closefds) < 0) {
 			dropbear_exit("Failed to daemonize: %s", strerror(errno));
 		}
 	}
