diff --git a/gensignkey.c b/gensignkey.c
index 8317fea..7c212ce 100644
--- a/gensignkey.c
+++ b/gensignkey.c
@@ -140,16 +140,13 @@ int signkey_generate(enum signkey_type keytype, int bits, const char* filename,
 		goto out;
 	}
 
-	if (link(fn_temp, filename) < 0) {
-		/* If generating keys on connection (skipexist) it's OK to get EEXIST 
-		- we probably just lost a race with another connection to generate the key */
-		if (!(skip_exist && errno == EEXIST)) {
-			dropbear_log(LOG_ERR, "Failed moving key file to %s: %s", filename,
-				strerror(errno));
-			/* XXX fallback to non-atomic copy for some filesystems? */
-			ret = DROPBEAR_FAILURE;
-			goto out;
-		}
+	/* Hack. We can't do hardlinks on vfat, so just use rename */
+	if (skip_exist && access(filename, F_OK) == 0) {
+		/* Ok. */
+	} else if (rename(fn_temp, filename) < 0) {
+		dropbear_log(LOG_ERR, "Failed moving key file to %s: %s", filename,
+		strerror(errno));
+		ret = DROPBEAR_FAILURE;
 	}
 
 out:
