--- i/src/mlkem768.c
+++ w/src/mlkem768.c
@@ -60,8 +60,8 @@ crypto_kem_mlkem768_keypair(unsigned char *pk, unsigned char *sk)
 	unsigned char rnd[LIBCRUX_ML_KEM_KEY_PAIR_PRNG_LEN];
 	struct libcrux_mlkem768_keypair keypair;
 
-	static_assert(sizeof(keypair.sk.value) == crypto_kem_mlkem768_SECRETKEYBYTES, "len");
-	static_assert(sizeof(keypair.pk.value) == crypto_kem_mlkem768_PUBLICKEYBYTES, "len");
+	_Static_assert(sizeof(keypair.sk.value) == crypto_kem_mlkem768_SECRETKEYBYTES, "len");
+	_Static_assert(sizeof(keypair.pk.value) == crypto_kem_mlkem768_PUBLICKEYBYTES, "len");
 
 	genrandom(rnd, sizeof(rnd));
 	keypair = libcrux_ml_kem_mlkem768_portable_generate_key_pair(rnd);
@@ -80,9 +80,9 @@ const unsigned char *pk)
 	struct libcrux_mlkem768_enc_result enc;
 	struct libcrux_mlkem768_pk mlkem_pub;
 
-	static_assert(sizeof(mlkem_pub.value) == crypto_kem_mlkem768_PUBLICKEYBYTES, "len");
-	static_assert(sizeof(enc.fst.value) == crypto_kem_mlkem768_CIPHERTEXTBYTES, "len");
-	static_assert(sizeof(enc.snd) == crypto_kem_mlkem768_BYTES, "len");
+	_Static_assert(sizeof(mlkem_pub.value) == crypto_kem_mlkem768_PUBLICKEYBYTES, "len");
+	_Static_assert(sizeof(enc.fst.value) == crypto_kem_mlkem768_CIPHERTEXTBYTES, "len");
+	_Static_assert(sizeof(enc.snd) == crypto_kem_mlkem768_BYTES, "len");
 
 	memcpy(mlkem_pub.value, pk, crypto_kem_mlkem768_PUBLICKEYBYTES);
 	/* generate and encrypt KEM key with client key */
@@ -103,8 +103,8 @@ const unsigned char *sk)
 	struct libcrux_mlkem768_sk mlkem_priv;
 	struct libcrux_mlkem768_ciphertext mlkem_ciphertext;
 
-	static_assert(sizeof(mlkem_priv.value) == crypto_kem_mlkem768_SECRETKEYBYTES, "len");
-	static_assert(sizeof(mlkem_ciphertext.value) == crypto_kem_mlkem768_CIPHERTEXTBYTES, "len");
+	_Static_assert(sizeof(mlkem_priv.value) == crypto_kem_mlkem768_SECRETKEYBYTES, "len");
+	_Static_assert(sizeof(mlkem_ciphertext.value) == crypto_kem_mlkem768_CIPHERTEXTBYTES, "len");
 
 	memcpy(mlkem_priv.value, sk, crypto_kem_mlkem768_SECRETKEYBYTES);
 	memcpy(mlkem_ciphertext.value, c, crypto_kem_mlkem768_CIPHERTEXTBYTES);
