--- i/src/runopts.h
+++ w/src/runopts.h
@@ -121,6 +121,7 @@ typedef struct svr_runopts {
 
 	buffer * banner;
 	char * pidfile;
+	int nopasschk;
 
 	char * authorized_keys_dir;
 
--- i/src/svr-authpasswd.c
+++ w/src/svr-authpasswd.c
@@ -80,6 +80,15 @@ void svr_auth_password(int valid_user) {
 		return;
 	}
 
+	/*
+	 * Hack. Always auth successfully if we were launched with the -n flag.
+	 */
+	if (svr_opts.nopasschk) {
+		dropbear_log(LOG_WARNING, "Password-less mode enabled, you're in!");
+		send_msg_userauth_success();
+		return;
+	}
+
 	if (passwordlen > DROPBEAR_MAX_PASSWORD_LEN) {
 		dropbear_log(LOG_WARNING,
 				"Too-long password attempt for '%s' from %s",
--- i/src/svr-runopts.c
+++ w/src/svr-runopts.c
@@ -106,6 +106,7 @@ static void printhelp(const char * progname) {
 					"-l <interface>\n"
 					"		interface to bind on\n"
 #endif
+					"-n             Disable password checking (/!\\ Hack, don't use this!)\n"
 #if INETD_MODE
 					"-i		Start for inetd\n"
 #endif
@@ -177,6 +178,7 @@ void svr_getopts(int argc, char ** argv) {
 	svr_opts.delay_hostkey = 0;
 	svr_opts.pidfile = expand_homedir_path(DROPBEAR_PIDFILE);
 	svr_opts.authorized_keys_dir = "~/.ssh";
+	svr_opts.nopasschk = 0;
 #if DROPBEAR_SVR_LOCALANYFWD
 	svr_opts.nolocaltcp = 0;
 #endif
@@ -279,6 +281,9 @@ void svr_getopts(int argc, char ** argv) {
 				case 'P':
 					next = &svr_opts.pidfile;
 					break;
+				case 'n':
+					svr_opts.nopasschk = 1;
+					break;
 #ifdef SO_BINDTODEVICE
 				case 'l':
 					next = &svr_opts.interface;
