--- i/Makefile.in
+++ w/Makefile.in
@@ -166,7 +166,7 @@
 	mkdir -p $@
 
 $(OBJ_DIR)/%.o: $(srcdir)/%.c $(HEADERS) | $(OBJ_DIR)
-	$(CC) $(CFLAGS) $(CPPFLAGS) $< -o $@ -c
+	$(if $V,,@echo "   * ${CC} $@"; )$(CC) $(CFLAGS) $(CPPFLAGS) $< -o $@ -c
 
 fuzz/%.o: $(srcdir)/../fuzz/%.c $(HEADERS)
 	$(CC) $(CFLAGS) $(CPPFLAGS) $< -o $@ -c
--- i/Makefile.in
+++ w/Makefile.in
--- i/libtomcrypt/Makefile.in
+++ w/libtomcrypt/Makefile.in
@@ -243,7 +243,7 @@
 #This is necessary for compatibility with BSD make (namely on OpenBSD)
 .SUFFIXES: .o .c
 .c.o:
-	$(CC) $(LTC_CFLAGS) -c $< -o $@
+	$(if $V,,@echo "   * ${CC} $@"; )$(CC) $(LTC_CFLAGS) -c $< -o $@
 
 #Create libtomcrypt.a
 $(LIBMAIN_S): $(OBJECTS)
--- i/libtommath/Makefile.in
+++ w/libtommath/Makefile.in
@@ -10,8 +10,6 @@
 CFLAGS += -Wno-deprecated
 CFLAGS += $(CPPFLAGS)
 
-V = 1
-
 ifeq ($V,1)
 silent=
 else
