diff --git a/c/configure.ac b/c/configure.ac
index 4b7566b..c745218 100644
--- a/c/configure.ac
+++ b/c/configure.ac
@@ -35,7 +35,7 @@ AC_TYPE_SIZE_T
 AC_CHECK_FUNCS(memcpy pwrite pread mkstemp)
 
 X_TYPE_SOCKLEN_T
-X_TYPE_IN_PORT_T
+#X_TYPE_IN_PORT_T
 X_DECL_H_ERRNO
 
 dnl Solaris needs -lsocket - and we need this for the getaddrinfo test
diff --git a/c/libzsync/sha1.h b/c/libzsync/sha1.h
index 3946505..c898185 100644
--- a/c/libzsync/sha1.h
+++ b/c/libzsync/sha1.h
@@ -10,6 +10,7 @@
 #define _SHA1_H
 
 #include "config.h"
+#include <stdio.h>
 
 #ifdef HAVE_INTTYPES_H
 #include <inttypes.h>
