diff --git a/c/libzsync/zsync.c b/c/libzsync/zsync.c
index 6a7486c..99f4381 100644
--- a/c/libzsync/zsync.c
+++ b/c/libzsync/zsync.c
@@ -41,6 +41,8 @@
 #include <string.h>
 #include <ctype.h>
 #include <time.h>
+#include <errno.h>
+#include <fcntl.h>
 
 #include <arpa/inet.h>
 
@@ -535,6 +537,17 @@ int zsync_rename_file(struct zsync_state *zs, const char *f) {
         free(rf);
         zs->cur_filename = strdup(f);
     }
+    else if (errno == EXDEV) {
+        int in_fd = open(rf, O_RDONLY);
+        int out_fd = open(f, O_WRONLY);
+        char buf[8192];
+        while(1) {
+            ssize_t result = read(in_fd, &buf[0], sizeof(buf));
+            if (!result) break;
+            write(out_fd, &buf[0], result);
+        }
+        unlink(rf);
+    }
     else
         perror("rename");
 
