diff -Nuarp sqlite-autoconf-3410000-orig/sqlite3.c sqlite-autoconf-3410000/sqlite3.c
--- sqlite-autoconf-3410000-orig/sqlite3.c	2023-02-21 20:38:33.000000000 +0100
+++ sqlite-autoconf-3410000/sqlite3.c	2023-02-28 17:55:04.580598850 +0100
@@ -126882,6 +126882,17 @@ static double xCeil(double x){ return ce
 static double xFloor(double x){ return floor(x); }
 
 /*
+** Some systems do not have log2() and log10() in their standard math
+** libraries.
+*/
+#if defined(HAVE_LOG10) && HAVE_LOG10==0
+# define log10(X) (0.4342944819032517867*log(X))
+#endif
+#if defined(HAVE_LOG2) && HAVE_LOG2==0
+# define log2(X) (1.442695040888963456*log(X))
+#endif
+
+/*
 ** Implementation of SQL functions:
 **
 **   ln(X)       - natural logarithm
