--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -8,8 +8,7 @@
 #
 # ##############################################################################
 
-cmake_minimum_required(VERSION 3.1.3)
-cmake_policy(SET CMP0054 NEW)
+cmake_minimum_required(VERSION 3.15)
 
 # In-source builds are disabled.
 if("${CMAKE_CURRENT_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_BINARY_DIR}")
@@ -43,8 +42,8 @@ project(
   leptonica
   LANGUAGES C
   VERSION 1.84.1)
-set(CMAKE_C_STANDARD 17)
+set(CMAKE_C_STANDARD 11)
 set(CMAKE_C_STANDARD_REQUIRED ON)
 
 # Standard installation paths
