diff --git a/src/utils.c b/src/utils.c
index bb3507c..9caf08f 100644
--- a/src/utils.c
+++ b/src/utils.c
@@ -1903,6 +1903,7 @@ FILE  *fp;
         return (FILE *)ERROR_PTR("data not defined", procName, NULL);
 
 #if HAVE_FMEMOPEN
+    __asm__(".symver fmemopen,fmemopen@GLIBC_2.4");
     if ((fp = fmemopen((void *)data, size, "rb")) == NULL)
         return (FILE *)ERROR_PTR("stream not opened", procName, NULL);
 #else  /* write to tmp file */
