diff --git a/src/usocket.c b/src/usocket.c
--- a/src/usocket.c
+++ b/src/usocket.c
@@ -18,11 +18,7 @@
 * Wait for readable/writable/connected socket with timeout
 \*-------------------------------------------------------------------------*/
 #ifndef SOCKET_SELECT
-#include <sys/poll.h>
 
-#define WAITFD_R        POLLIN
-#define WAITFD_W        POLLOUT
-#define WAITFD_C        (POLLIN|POLLOUT)
 int socket_waitfd(p_socket ps, int sw, p_timeout tm) {
     int ret;
     struct pollfd pfd;
@@ -41,10 +37,6 @@ int socket_waitfd(p_socket ps, int sw, p_timeout tm) {
 }
 #else
 
-#define WAITFD_R        1
-#define WAITFD_W        2
-#define WAITFD_C        (WAITFD_R|WAITFD_W)
-
 int socket_waitfd(p_socket ps, int sw, p_timeout tm) {
     int ret;
     fd_set rfds, wfds, *rp, *wp;
diff --git a/src/usocket.h b/src/usocket.h
--- a/src/usocket.h
+++ b/src/usocket.h
@@ -56,4 +56,15 @@ typedef struct sockaddr_storage t_sockaddr_storage;
 
 #define SOCKET_INVALID (-1)
 
+#ifndef SOCKET_SELECT
+#include <sys/poll.h>
+#define WAITFD_R        POLLIN
+#define WAITFD_W        POLLOUT
+#define WAITFD_C        (POLLIN|POLLOUT)
+#else
+#define WAITFD_R        1
+#define WAITFD_W        2
+#define WAITFD_C        (WAITFD_R|WAITFD_W)
+#endif
+
 #endif /* USOCKET_H */
diff --git a/src/wsocket.c b/src/wsocket.c
--- a/src/wsocket.c
+++ b/src/wsocket.c
@@ -42,11 +42,6 @@ int socket_close(void) {
 /*-------------------------------------------------------------------------*\
 * Wait for readable/writable/connected socket with timeout
 \*-------------------------------------------------------------------------*/
-#define WAITFD_R        1
-#define WAITFD_W        2
-#define WAITFD_E        4
-#define WAITFD_C        (WAITFD_E|WAITFD_W)
-
 int socket_waitfd(p_socket ps, int sw, p_timeout tm) {
     int ret;
     fd_set rfds, wfds, efds, *rp = NULL, *wp = NULL, *ep = NULL;
diff --git a/src/wsocket.h b/src/wsocket.h
--- a/src/wsocket.h
+++ b/src/wsocket.h
@@ -16,6 +16,11 @@ typedef SOCKADDR_STORAGE t_sockaddr_storage;
 typedef SOCKET t_socket;
 typedef t_socket *p_socket;
 
+#define WAITFD_R        1
+#define WAITFD_W        2
+#define WAITFD_E        4
+#define WAITFD_C        (WAITFD_E|WAITFD_W)
+
 #ifndef IPV6_V6ONLY
 #define IPV6_V6ONLY 27
 #endif
