diff --git a/src/usocket.c b/src/usocket.c
index 7965db6..fc1f607 100644
--- a/src/usocket.c
+++ b/src/usocket.c
@@ -122,8 +122,15 @@ int socket_select(t_socket n, fd_set *rfds, fd_set *wfds, fd_set *efds,
 \*-------------------------------------------------------------------------*/
 int socket_create(p_socket ps, int domain, int type, int protocol) {
     *ps = socket(domain, type, protocol);
-    if (*ps != SOCKET_INVALID) return IO_DONE;
-    else return errno;
+    if (*ps != SOCKET_INVALID) {
+        // Enforce CLOEXEC
+        int flags = fcntl(*ps, F_GETFD, 0);
+        flags |= O_CLOEXEC;
+        fcntl(*ps, F_SETFD, flags);
+        return IO_DONE;
+    } else {
+        return errno;
+    }
 }
 
 /*-------------------------------------------------------------------------*\
