--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -6,7 +6,7 @@
 ########################################################################
 # Project setup
 ########################################################################
-cmake_minimum_required(VERSION 2.8.12)
+cmake_minimum_required(VERSION 3.15)
 project(czmq)
 enable_language(C)
 enable_testing()
@@ -123,6 +123,13 @@
     set(MORE_LIBRARIES)
 endif()
 
+set(CMAKE_REQUIRED_LIBRARIES -lrt)
+check_symbol_exists(clock_gettime "time.h" HAVE_CLOCK_GETTIME)
+set(CMAKE_REQUIRED_LIBRARIES)
+if(HAVE_CLOCK_GETTIME)
+    list(APPEND MORE_LIBRARIES -lrt)
+endif()
+
 list(APPEND CMAKE_MODULE_PATH "${SOURCE_DIR}")
 set(OPTIONAL_LIBRARIES)
 set(OPTIONAL_LIBRARIES_STATIC)
@@ -544,6 +549,7 @@
 ########################################################################
 # executables
 ########################################################################
+if(FALSE)
 add_executable(
     zmakecert
     "${SOURCE_DIR}/src/zmakecert.c"
@@ -631,6 +637,7 @@
     ${OPTIONAL_LIBRARIES_STATIC}
 )
 endif()
+endif()
 
 ########################################################################
 # tests
