diff --git a/include/czmq_prelude.h b/include/czmq_prelude.h
index c77e922..890f743 100644
--- a/include/czmq_prelude.h
+++ b/include/czmq_prelude.h
@@ -155,6 +155,9 @@
 #   ifndef _BSD_SOURCE
 #   define _BSD_SOURCE                  //  Include stuff from 4.3 BSD Unix
 #   endif
+#   ifndef _DEFAULT_SOURCE
+#   define _DEFAULT_SOURCE
+#   endif
 #elif (defined (Mips))
 #   define __UTYPE_MIPS
 #   define __UNIX__
