diff --git a/Makefile b/Makefile
index 114a6eb6..818d4ce5 100644
--- a/Makefile
+++ b/Makefile
@@ -334,7 +334,6 @@ THIRD_OBJ := \
 	$(HARFBUZZ_OBJ) \
 	$(JBIG2DEC_OBJ) \
 	$(JPEGXR_OBJ) \
-	$(LIBJPEG_OBJ) \
 	$(LURATECH_OBJ) \
 	$(MUJS_OBJ) \
 	$(OPENJPEG_OBJ) \
diff --git a/Makethird b/Makethird
index fbc8b7f8..47b4ba1b 100644
--- a/Makethird
+++ b/Makethird
@@ -459,9 +459,9 @@ $(LIBJPEG_OUT):
 $(LIBJPEG_OUT)/%.o: $(LIBJPEG_DIR)/%.c | $(LIBJPEG_OUT)
 	$(CC_CMD) -Dmain=xxxmain
 
-LIBJPEG_CFLAGS := -Iscripts/libjpeg -I$(LIBJPEG_DIR)
+LIBJPEG_CFLAGS := -Iscripts/libjpeg -I$(LIBJPEG_DIR) -DSHARE_JPEG=1
 else
-LIBJPEG_CFLAGS := $(SYS_LIBJPEG_CFLAGS) -DSHARE_JPEG
+LIBJPEG_CFLAGS := $(SYS_LIBJPEG_CFLAGS) -DSHARE_JPEG=1
 LIBJPEG_LIBS := $(SYS_LIBJPEG_LIBS)
 endif
 
