diff --git a/source/fitz/load-png.c b/source/fitz/load-png.c
index daa0814a..636447d4 100644
--- a/source/fitz/load-png.c
+++ b/source/fitz/load-png.c
@@ -358,7 +358,7 @@ png_read_icc(fz_context *ctx, struct info *info, const unsigned char *p, unsigne
 	fz_always(ctx)
 		fz_drop_stream(ctx, stm);
 	fz_catch(ctx)
-		fz_rethrow(ctx);
+		fz_warn(ctx, "could not load ICC profile in PNG image");
 
 	/* drop old one in case we have multiple ICC profiles */
 	fz_drop_colorspace(ctx, info->cs);
