diff --git a/Makefile b/Makefile
index 4a6fb4ffb..df6a44292 100644
--- a/Makefile
+++ b/Makefile
@@ -448,6 +448,8 @@ install-headers:
 	install -d $(DESTDIR)$(incdir)/mupdf/pdf
 	install -m 644 include/mupdf/*.h $(DESTDIR)$(incdir)/mupdf
 	install -m 644 include/mupdf/fitz/*.h $(DESTDIR)$(incdir)/mupdf/fitz
+	install -m 644 source/fitz/xml-imp.h $(DESTDIR)$(incdir)/mupdf/fitz
+	install -m 644 source/html/html-imp.h $(DESTDIR)$(incdir)/mupdf/fitz
 	install -m 644 include/mupdf/pdf/*.h $(DESTDIR)$(incdir)/mupdf/pdf

 install-libs: libs install-headers

diff --git a/source/html/html-imp.h b/source/html/html-imp.h
index c47aef640..e8c4fe0f1 100644
--- a/source/html/html-imp.h
+++ b/source/html/html-imp.h
@@ -516,17 +516,17 @@ struct fz_html_flow_s
 };


-fz_css *fz_new_css(fz_context *ctx);
-void fz_parse_css(fz_context *ctx, fz_css *css, const char *source, const char *file);
+FZ_FUNCTION fz_css *fz_new_css(fz_context *ctx);
+FZ_FUNCTION void fz_parse_css(fz_context *ctx, fz_css *css, const char *source, const char *file);
 fz_css_property *fz_parse_css_properties(fz_context *ctx, fz_pool *pool, const char *source);
-void fz_drop_css(fz_context *ctx, fz_css *css);
+FZ_FUNCTION void fz_drop_css(fz_context *ctx, fz_css *css);
 void fz_debug_css(fz_context *ctx, fz_css *css);
 const char *fz_css_property_name(int name);

-void fz_match_css(fz_context *ctx, fz_css_match *match, fz_css_match *up, fz_css *css, fz_xml *node);
+FZ_FUNCTION void fz_match_css(fz_context *ctx, fz_css_match *match, fz_css_match *up, fz_css *css, fz_xml *node);
 void fz_match_css_at_page(fz_context *ctx, fz_css_match *match, fz_css *css);

-int fz_get_css_match_display(fz_css_match *node);
+FZ_FUNCTION int fz_get_css_match_display(fz_css_match *node);
 void fz_default_css_style(fz_context *ctx, fz_css_style *style);
 void fz_apply_css_style(fz_context *ctx, fz_html_font_set *set, fz_css_style *style, fz_css_match *match);
