diff --git a/Makefile b/Makefile
index 1e760280f..10732a6a7 100644
--- a/Makefile
+++ b/Makefile
@@ -299,7 +299,7 @@ generate: source/html/css-properties.h
 # --- Library ---
 
 ifeq ($(shared),yes)
-  $(OUT)/libmupdf.$(SO)$(SO_VERSION): $(MUPDF_OBJ) $(THIRD_OBJ)
+  $(OUT)/libmupdf.$(SO)$(SO_VERSION): $(MUPDF_OBJ) $(THIRD_OBJ) $(XLIBS)
 	$(LINK_SO_CMD) $(THIRD_LIBS) $(LIBCRYPTO_LIBS) $(LIBS)
   ifeq ($(OS),OpenBSD)
     # should never create symlink
