diff --git a/Makefile b/Makefile
index e806b39a..571f61aa 100644
--- a/Makefile
+++ b/Makefile
@@ -102,7 +102,7 @@ $(OUT)/%.a :
 $(OUT)/%.exe: %.c
 	$(LINK_CMD)
 
-$(OUT)/%.$(SO)$(SO_VERSION):
+$(OUT)/%.$(SO)$(SO_VERSION): $(XLIBS)
 ifeq ($(SO_VERSION_LINUX),yes)
 	$(LINK_CMD) -Wl,-soname,$(notdir $@) $(LIB_LDFLAGS) $(THIRD_LIBS) $(LIBCRYPTO_LIBS)
 	ln -sf $(notdir $@) $(patsubst %$(SO_VERSION), %, $@)
