diff --git a/Makefile b/Makefile
index d247334bd..11c2918ea 100644
--- a/Makefile
+++ b/Makefile
@@ -293,7 +293,7 @@ generate: source/html/css-properties.h
 # --- Library ---
 
 ifeq ($(shared),yes)
-  $(OUT)/libmupdf.$(SO)$(SO_VERSION): $(MUPDF_OBJ) $(THIRD_OBJ)
+  $(OUT)/libmupdf.$(SO)$(SO_VERSION): $(MUPDF_OBJ) $(THIRD_OBJ) $(XLIBS)
 	$(LINK_SO_CMD) $(THIRD_LIBS) $(LIBCRYPTO_LIBS)
   ifeq ($(OS),OpenBSD)
     # should never create symlink
