--- i/lib/CMakeLists.txt
+++ w/lib/CMakeLists.txt
@@ -21,8 +21,8 @@
 # SPDX-License-Identifier: curl
 #
 ###########################################################################
-set(LIB_NAME libcurl)
-set(LIBCURL_OUTPUT_NAME libcurl CACHE STRING "Basename of the curl library")
+set(LIB_NAME libcurl_ko)
+set(LIBCURL_OUTPUT_NAME libcurl_ko CACHE STRING "Basename of the curl library")
 add_definitions(-DBUILDING_LIBCURL)
 
 configure_file(curl_config.h.cmake
--- i/libcurl.pc.in
+++ w/libcurl.pc.in
@@ -36,6 +36,6 @@ Name: libcurl
 URL: https://curl.se/
 Description: Library to transfer files with ftp, http, etc.
 Version: @CURLVERSION@
-Libs: -L${libdir} -lcurl @LIBCURL_NO_SHARED@
+Libs: -L${libdir} -lcurl_ko @LIBCURL_NO_SHARED@
 Libs.private: @LIBCURL_LIBS@
 Cflags: -I${includedir} @CPPFLAG_CURL_STATICLIB@
