diff -Nuarp tar-1.33-ori/gnu/fchmodat.c tar-1.33/gnu/fchmodat.c
--- tar-1.33-ori/gnu/fchmodat.c	2021-01-07 15:29:44.000000000 +0100
+++ tar-1.33/gnu/fchmodat.c	2021-01-16 00:41:39.414012004 +0100
@@ -68,7 +68,7 @@ fchmodat (int dir, char const *file, mod
     {
       struct stat st;
 
-#  if defined O_PATH && defined AT_EMPTY_PATH
+#  if defined NOPE_PATH && defined AT_EMPTY_PATH
       /* Open a file descriptor with O_NOFOLLOW, to make sure we don't
          follow symbolic links, if /proc is mounted.  O_PATH is used to
          avoid a failure if the file is not readable.
diff -Nuarp tar-1.33-ori/gnu/lchmod.c tar-1.33/gnu/lchmod.c
--- tar-1.33-ori/gnu/lchmod.c	2021-01-07 15:28:45.000000000 +0100
+++ tar-1.33/gnu/lchmod.c	2021-01-16 00:41:28.708012124 +0100
@@ -45,7 +45,7 @@
 int
 lchmod (char const *file, mode_t mode)
 {
-#if defined O_PATH && defined AT_EMPTY_PATH
+#if defined NOPE_PATH && defined AT_EMPTY_PATH
   /* Open a file descriptor with O_NOFOLLOW, to make sure we don't
      follow symbolic links, if /proc is mounted.  O_PATH is used to
      avoid a failure if the file is not readable.
