--- i/CMakeLists.txt
+++ w/CMakeLists.txt
@@ -1,6 +1,6 @@
 # CMake build script for ZeroMQ
 
-cmake_minimum_required(VERSION 2.8)
+cmake_minimum_required(VERSION 3.15)
 project(ZeroMQ)
 
 option(WITH_OPENPGM "Build with support for OpenPGM" OFF)
@@ -266,10 +266,6 @@ elseif(CMAKE_SYSTEM_NAME MATCHES "Darwin")
   add_definitions(-D_DARWIN_C_SOURCE)
 endif()
 
-set(CMAKE_PYTHON_VERSION 2.7 2.6 2.5 2.4)
-find_package(PythonInterp)
-find_package(AsciiDoc)
-
 cmake_dependent_option(WITH_DOC "Build Reference Guide documentation(requires DocBook)" ON
                        "PYTHON_FOUND;ASCIIDOC_FOUND" OFF)
 
@@ -609,6 +605,7 @@ else()
     else()
       set_target_properties(libzmq PROPERTIES
                              OUTPUT_NAME "zmq"
+                             VERSION 4
                              PUBLIC_HEADER "${public_headers}")
     endif()
 endif()
