--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -115,7 +115,6 @@
 check_include_files(ifaddrs.h ZMQ_HAVE_IFADDRS)
 check_include_files(windows.h ZMQ_HAVE_WINDOWS)
 check_include_files(sys/uio.h ZMQ_HAVE_UIO)
-check_include_files(sys/eventfd.h ZMQ_HAVE_EVENTFD)
 
 check_library_exists(ws2_32 printf "" HAVE_WS2_32) # TODO: Why doesn't something logical like WSAStartup work?
 check_library_exists(ws2 printf "" HAVE_WS2)
