diff -uNr a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt	2020-06-07 17:18:56.373394505 +0200
+++ b/CMakeLists.txt	2020-06-07 21:42:18.129006112 +0200
@@ -31,8 +31,12 @@
         set(PLAT "linux")
         set(LINK_FLAGS "-shared")
     endif(APPLE)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -Wall -fPIC -march=native")
-    set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -Wall -fPIC -march=native")
+    add_compile_options(-g -Wall -fPIC)
+    include(CheckCXXCompilerFlag)
+    CHECK_CXX_COMPILER_FLAG("-march=native" COMPILER_SUPPORTS_ARCH_NATIVE)
+    if (COMPILER_SUPPORTS_ARCH_NATIVE)
+        add_compile_options(-march=native)
+    endif()
 else(UNIX)
     if(WIN32)
         set(PLAT "win32")
