diff --git a/CMakeLists.txt b/CMakeLists.txt
index ddc8716..b92767d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -46,6 +46,9 @@ else(UNIX)
     endif(WIN32)
 endif(UNIX)
 
+set(CMAKE_BUILD_RPATH "$ORIGIN/../../../../libs")
+set(CMAKE_INSTALL_RPATH "$ORIGIN/../../../../libs")
+
 if(NOT LUA_INCLUDE_DIR OR (WIN32 AND NOT LUA_LIBRARIES))
     set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/cmake/Modules/")
     find_package(Lua REQUIRED)
