diff --git a/libdjvu/ByteStream.cpp b/libdjvu/ByteStream.cpp
index 840885d..4162346 100644
--- a/libdjvu/ByteStream.cpp
+++ b/libdjvu/ByteStream.cpp
@@ -671,22 +671,23 @@ urlfopen(const GURL &url,const char mode[])
        }
   }
   return retval?retval:fopen((const char *)url.NativeFilename(),mode);
-#elif defined(__APPLE__)
-  return fopen((const char *)url.UTF8Filename(),mode);
-#else
-  return fopen((const char *)url.NativeFilename(),mode);
 #endif
+//#elif defined(__APPLE__)
+  return fopen((const char *)url.UTF8Filename(),mode);
+//#else
+  //return fopen((const char *)url.NativeFilename(),mode);
+//#endif
 }
 
 #ifdef UNIX
 static int
 urlopen(const GURL &url, const int mode, const int perm)
 {
-#if defined(__APPLE__)
+//#if defined(__APPLE__)
   return open((const char *)url.UTF8Filename(),mode,perm);
-#else
-  return open((const char *)url.NativeFilename(),mode,perm);
-#endif
+//#else
+  //return open((const char *)url.NativeFilename(),mode,perm);
+//#endif
 }
 #endif /* UNIX */
