diff -NuarpZ gettext-0.21-ori/libtextstyle/lib/glib/gstrfuncs.c gettext-0.21/libtextstyle/lib/glib/gstrfuncs.c
--- gettext-0.21-ori/libtextstyle/lib/glib/gstrfuncs.c	2020-04-11 12:01:49.000000000 +0200
+++ gettext-0.21/libtextstyle/lib/glib/gstrfuncs.c	2021-02-15 16:10:30.130191728 +0100
@@ -555,9 +555,6 @@ g_ascii_formatd (gchar       *buffer,
 		 const gchar *format,
 		 gdouble      d)
 {
-  struct lconv *locale_data;
-  const char *decimal_point;
-  int decimal_point_len;
   gchar *p;
   int rest_len;
   gchar format_char;
@@ -587,9 +584,16 @@ g_ascii_formatd (gchar       *buffer,
       
   _g_snprintf (buffer, buf_len, format, d);
 
+#if defined(__ANDROID__) && __ANDROID_API__ < 21
+  const char decimal_point[] = ".";
+  int decimal_point_len = 1;
+#else
+  struct lconv *locale_data;
+  const char *decimal_point;
   locale_data = localeconv ();
   decimal_point = locale_data->decimal_point;
-  decimal_point_len = strlen (decimal_point);
+  int decimal_point_len = strlen (decimal_point);
+#endif
 
   g_assert (decimal_point_len != 0);
 
