diff --git a/CREDITS b/CREDITS
new file mode 100644
index 0000000..992fb93
--- /dev/null
+++ b/CREDITS
@@ -0,0 +1 @@
+Taken from http://code.google.com/p/popen-noshell/
diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..069a7bb
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,17 @@
+SRCS=popen_noshell.c
+
+OBJS:=$(SRCS:%.c=%.o)
+
+%.o: %.c
+	$(CC) $(CPPFLAGS) $(CFLAGS) -o $@ -c $<
+
+all: libpopen_noshell.a
+
+libpopen_noshell.a: $(OBJS)
+	$(AR) rcs $@ $(OBJS)
+
+clean:
+	rm -rf *.o
+	rm -rf libpopen_noshell.a
+
+.PHONY: clean
