diff --git a/Makefile b/Makefile
index 76e8a3a..acdd7c9 100644
--- a/Makefile
+++ b/Makefile
@@ -37,17 +37,18 @@ DEBUG_CFLAGS:=-Og -fno-omit-frame-pointer -pipe -g
 OPT_CFLAGS:=-O2 -fomit-frame-pointer -pipe
 
 ifdef DEBUG
-	OUT_DIR:=Debug
+	OUT_DIR?=Debug
 	CFLAGS?=$(DEBUG_CFLAGS)
 	CXXFLAGS?=$(DEBUG_CFLAGS)
 	EXTRA_CPPFLAGS+=-DDEBUG
 	STRIP:=true
 else
-	OUT_DIR:=Release
+	OUT_DIR?=Release
 	CFLAGS?=$(OPT_CFLAGS)
 	CXXFLAGS?=$(OPT_CFLAGS)
 	EXTRA_CPPFLAGS+=-DNDEBUG
 endif
+export OUT_DIR
 
 # Explictly enforce debug CFLAGS for the debug target (vs., simply having DEBUG set in the env)
 ifdef DEBUGFLAGS
@@ -335,11 +336,7 @@ endif
 LDFLAGS?=-Wl,--as-needed
 
 # And we want to link against our own library ;).
-ifdef DEBUG
-	EXTRA_LDFLAGS+=-L./Debug
-else
-	EXTRA_LDFLAGS+=-L./Release
-endif
+EXTRA_LDFLAGS+=-L$(OUT_DIR)
 ifdef SHARED
 	LIBS+=-lfbink
 else
@@ -348,23 +345,23 @@ endif
 
 # Pick up our vendored build of libunibreak, if requested
 ifdef UNIBREAK
-	EXTRA_LDFLAGS+=-Llibunibreak-staged/src/.libs
+	EXTRA_LDFLAGS+=-L$(OUT_DIR)/libunibreak-staged/src/.libs
 	LIBS+=-l:libunibreak.a
 	SHARED_LIBS+=-l:libunibreak.a
 endif
 
 # Same for libi2c, on Kobo
 ifdef KOBO
-	EXTRA_CPPFLAGS+=-Ilibi2c-staged/include
-	EXTRA_LDFLAGS+=-Llibi2c-staged/lib
+	EXTRA_CPPFLAGS+=-I$(OUT_DIR)/libi2c-staged/include
+	EXTRA_LDFLAGS+=-L$(OUT_DIR)/libi2c-staged/lib
 	I2C_LIBS:=-l:libi2c.a
 	LIBS+=$(I2C_LIBS)
 	SHARED_LIBS+=$(I2C_LIBS)
 endif
 
 # Pick up our vendored build of libevdev (for the PoC that relies on it)
-EVDEV_CPPFLAGS:=-Ilibevdev-staged/include/libevdev-1.0
-EVDEV_LDFLAGS:=-Llibevdev-staged/lib
+EVDEV_CPPFLAGS:=-I$(OUT_DIR)/libevdev-staged/include/libevdev-1.0
+EVDEV_LDFLAGS:=-L$(OUT_DIR)/libevdev-staged/lib
 EVDEV_LIBS:=-l:libevdev.a
 
 # And with our own rpath for standalone distribution
@@ -521,14 +518,14 @@ FBINK_STATIC_NAME:=libfbink.a
 
 default: all
 
-SHAREDLIB_OBJS:=$(addprefix $(OUT_DIR)/shared/, $(LIB_SRCS:.c=.o))
-UB_SHAREDLIB_OBJS:=$(addprefix $(OUT_DIR)/shared/, $(LIB_UB_SRCS:.c=.o))
-QT_SHAREDLIB_OBJS:=$(addprefix $(OUT_DIR)/shared/, $(LIB_QT_SRCS:.c=.o))
-STATICLIB_OBJS:=$(addprefix $(OUT_DIR)/static/, $(LIB_SRCS:.c=.o))
-UB_STATICLIB_OBJS:=$(addprefix $(OUT_DIR)/static/, $(LIB_UB_SRCS:.c=.o))
-QT_STATICLIB_OBJS:=$(addprefix $(OUT_DIR)/static/, $(LIB_QT_SRCS:.c=.o))
-CMD_OBJS:=$(addprefix $(OUT_DIR)/, $(CMD_SRCS:.c=.o))
-BTN_OBJS:=$(addprefix $(OUT_DIR)/, $(BTN_SRCS:.c=.o))
+SHAREDLIB_OBJS=$(addprefix $(OUT_DIR)/shared/, $(LIB_SRCS:.c=.o))
+UB_SHAREDLIB_OBJS=$(addprefix $(OUT_DIR)/shared/, $(LIB_UB_SRCS:.c=.o))
+QT_SHAREDLIB_OBJS=$(addprefix $(OUT_DIR)/shared/, $(LIB_QT_SRCS:.c=.o))
+STATICLIB_OBJS=$(addprefix $(OUT_DIR)/static/, $(LIB_SRCS:.c=.o))
+UB_STATICLIB_OBJS=$(addprefix $(OUT_DIR)/static/, $(LIB_UB_SRCS:.c=.o))
+QT_STATICLIB_OBJS=$(addprefix $(OUT_DIR)/static/, $(LIB_QT_SRCS:.c=.o))
+CMD_OBJS=$(addprefix $(OUT_DIR)/, $(CMD_SRCS:.c=.o))
+BTN_OBJS=$(addprefix $(OUT_DIR)/, $(BTN_SRCS:.c=.o))
 
 # Silence a few warnings, only when specifically compiling libunibreak...
 # c.f., https://stackoverflow.com/q/1305665
@@ -558,15 +555,17 @@ $(OUT_DIR)/static/%.o: %.c
 $(OUT_DIR)/%.o: %.c
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(FEATURES_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) -o $@ -c $<
 
-outdir:
-	mkdir -p $(OUT_DIR)/shared/cutef8 $(OUT_DIR)/static/cutef8 $(OUT_DIR)/shared/libunibreak/src $(OUT_DIR)/static/libunibreak/src $(OUT_DIR)/shared/qimagescale $(OUT_DIR)/static/qimagescale
+outdir: $(OUT_DIR)/shared/cutef8/ $(OUT_DIR)/static/cutef8/ $(OUT_DIR)/shared/libunibreak/src/ $(OUT_DIR)/static/libunibreak/src/ $(OUT_DIR)/shared/qimagescale/ $(OUT_DIR)/static/qimagescale/
+
+$(OUT_DIR)/%/:
+	mkdir -p $@
 
 # Make absolutely sure we create our output directories first, even with unfortunate // timings!
 # c.f., https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html#Prerequisite-Types
-$(SHAREDLIB_OBJS): libi2c.built | outdir
+$(SHAREDLIB_OBJS): $(OUT_DIR)/libi2c.built | outdir
 $(UB_SHAREDLIB_OBJS): | outdir
 $(QT_SHAREDLIB_OBJS): | outdir
-$(STATICLIB_OBJS): libi2c.built | outdir
+$(STATICLIB_OBJS): $(OUT_DIR)/libi2c.built | outdir
 $(UB_STATICLIB_OBJS): | outdir
 $(QT_STATICLIB_OBJS): | outdir
 $(CMD_OBJS): | outdir
@@ -575,11 +574,11 @@ $(BTN_OBJS): | outdir
 all: static
 
 ifdef UNIBREAK
-staticlib: $(STATICLIB_OBJS) libunibreak.built
+staticlib: $(STATICLIB_OBJS) $(OUT_DIR)/libunibreak.built
 	$(AR) $(FBINK_STATIC_FLAGS) $(OUT_DIR)/$(FBINK_STATIC_NAME) $(STATICLIB_OBJS)
 	$(RANLIB) $(OUT_DIR)/$(FBINK_STATIC_NAME)
 
-sharedlib: $(SHAREDLIB_OBJS) libunibreak.built
+sharedlib: $(SHAREDLIB_OBJS) $(OUT_DIR)/libunibreak.built
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(FEATURES_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LIB_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) $(FBINK_SHARED_FLAGS) -o$(OUT_DIR)/$(FBINK_SHARED_NAME_FILE) $(SHAREDLIB_OBJS)
 	ln -sf $(FBINK_SHARED_NAME_FILE) $(OUT_DIR)/$(FBINK_SHARED_NAME)
 	ln -sf $(FBINK_SHARED_NAME_FILE) $(OUT_DIR)/$(FBINK_SHARED_NAME_VER)
@@ -636,7 +635,7 @@ ifdef LINUX
 utils: | outdir
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(DOOM_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LIB_CFLAGS) $(LTO_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/doom utils/doom.c -lrt
 else
-utils: libi2c.built | outdir
+utils: $(OUT_DIR)/libi2c.built | outdir
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(TOOLS_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LIB_CFLAGS) $(LTO_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/rota utils/rota.c
 	$(STRIP) --strip-unneeded $(OUT_DIR)/rota
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(DOOM_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LIB_CFLAGS) $(LTO_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/doom utils/doom.c -lrt $(UTILS_LIBS) $(I2C_LIBS)
@@ -644,13 +643,13 @@ utils: libi2c.built | outdir
 endif
 
 ifdef KOBO
-alt: libi2c.built | outdir
+alt: $(OUT_DIR)/libi2c.built | outdir
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(DOOM_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LIB_CFLAGS) $(LTO_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/alt_buffer utils/alt_buffer.c $(I2C_LIBS)
 	$(STRIP) --strip-unneeded $(OUT_DIR)/alt_buffer
 endif
 
 ifdef KOBO
-sunxi: libi2c.built | outdir
+sunxi: $(OUT_DIR)/libi2c.built | outdir
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/ion_heaps utils/ion_heaps.c
 	$(STRIP) --strip-unneeded $(OUT_DIR)/ion_heaps
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) -o$(OUT_DIR)/kx122_i2c utils/kx122_i2c.c $(I2C_LIBS)
@@ -658,7 +657,7 @@ sunxi: libi2c.built | outdir
 endif
 
 ifdef KOBO
-ftrace: libevdev.built tiny | outdir
+ftrace: $(OUT_DIR)/libevdev.built tiny | outdir
 	$(CC) $(CPPFLAGS) $(EXTRA_CPPFLAGS) $(EVDEV_CPPFLAGS) $(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(LDFLAGS) $(EXTRA_LDFLAGS) $(EVDEV_LDFLAGS) -o$(OUT_DIR)/finger_trace utils/finger_trace.c $(LIBS) $(EVDEV_LIBS)
 	$(STRIP) --strip-unneeded $(OUT_DIR)/finger_trace
 endif
@@ -719,45 +718,46 @@ remarkable:
 pocketbook:
 	$(MAKE) strip POCKETBOOK=true
 
-libunibreak.built:
-	mkdir -p libunibreak-staged
+$(OUT_DIR)/libunibreak.built:
+	mkdir -p $(OUT_DIR)/libunibreak-staged
 	cd libunibreak && \
 	env NOCONFIGURE=1 ./autogen.sh
-	cd libunibreak-staged && \
+	cd $(OUT_DIR)/libunibreak-staged && \
 	env CPPFLAGS='$(CPPFLAGS) $(EXTRA_CPPFLAGS)' \
 	CFLAGS='$(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(UNIBREAK_CFLAGS)' \
 	LDFLAGS='$(LDFLAGS)' \
-	../libunibreak/configure \
+	"$$OLDPWD/libunibreak/configure" \
 	$(if $(CROSS_TC),--host=$(CROSS_TC),) \
 	--enable-static \
 	--disable-shared \
 	$(if $(SHARED),--with-pic=yes,)
-	$(MAKE) -C libunibreak-staged
-	touch libunibreak.built
+	$(MAKE) -C $(OUT_DIR)/libunibreak-staged
+	touch $@
 
-libi2c.built:
-	mkdir -p libi2c-staged
+$(OUT_DIR)/libi2c.built:
+	mkdir -p $(OUT_DIR)/libi2c-staged
 	$(MAKE) -C i2c-tools \
 	BUILD_DYNAMIC_LIB=0 USE_STATIC_LIB=1 BUILD_STATIC_LIB=1 V=1 \
 	CC='$(CC)' AR='$(AR)' \
 	CFLAGS='$(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(I2C_CFLAGS)' \
-	PREFIX="/" libdir="/lib" DESTDIR='$(CURDIR)/libi2c-staged' \
+	PREFIX="/" libdir="/lib" DESTDIR='$(abspath $(OUT_DIR)/libi2c-staged)' \
 	install-lib install-include
-	touch libi2c.built
+	touch $@
 
-libevdev.built:
-	mkdir -p libevdev-staged
+$(OUT_DIR)/libevdev.built:
+	mkdir -p $(addprefix $(OUT_DIR)/libevdev-,build staged)
 	cd libevdev && \
-	autoreconf -fi && \
+	autoreconf -fi
+	cd $(OUT_DIR)/libevdev-build &&
 	env CPPFLAGS='$(CPPFLAGS) $(EXTRA_CPPFLAGS)' \
 	CFLAGS='$(CFLAGS) $(EXTRA_CFLAGS) $(SHARED_CFLAGS) $(EVDEV_CFLAGS)' \
 	LDFLAGS='$(LDFLAGS)' \
-	./configure $(if $(CROSS_TC),--host=$(CROSS_TC),) \
-	--prefix='$(CURDIR)/libevdev-staged' \
+	"$$OLDPWD/libevdev/configure" $(if $(CROSS_TC),--host=$(CROSS_TC),) \
+	--prefix='$(abspath $(OUT_DIR)/libevdev-staged)' \
 	--enable-static \
 	--disable-shared && \
 	$(MAKE) install
-	touch libevdev.built
+	touch $@
 
 ifeq "$(CC_IS_CLANG)" "1"
 armcheck:
@@ -839,82 +839,48 @@ libevdevclean:
 	git clean -fxdq
 
 clean:
-	rm -rf Kobo/
-	rm -rf Release/*.a
-	rm -rf Release/*.so*
-	rm -rf Release/shared/*.o
-	rm -rf Release/shared/*.opt.yaml
-	rm -rf Release/shared/cutef8/*.o
-	rm -rf Release/shared/cutef8/*.opt.yaml
-	rm -rf Release/shared/libunibreak/src/*.o
-	rm -rf Release/shared/libunibreak/src/*.opt.yaml
-	rm -rf Release/shared/qimagescale/*.o
-	rm -rf Release/shared/qimagescale/*.opt.yaml
-	rm -rf Release/shared/utf8
-	rm -rf Release/static/*.o
-	rm -rf Release/static/*.opt.yaml
-	rm -rf Release/static/cutef8/*.o
-	rm -rf Release/static/cutef8/*.opt.yaml
-	rm -rf Release/static/libunibreak/src/*.o
-	rm -rf Release/static/libunibreak/src/*.opt.yaml
-	rm -rf Release/static/qimagescale/*.o
-	rm -rf Release/static/qimagescale/*.opt.yaml
-	rm -rf Release/static/utf8
-	rm -rf Release/*.o
-	rm -rf Release/*.opt.yaml
-	rm -rf Release/fbink
-	rm -rf Release/button_scan
-	rm -rf Release/rota
-	rm -rf Release/fbdepth
-	rm -rf Release/alt_buffer
-	rm -rf Release/doom
-	rm -rf Release/dump
-	rm -rf Release/Kobo-DevCap-Test.tar.gz
-	rm -rf Release/kx122_i2c
-	rm -rf Release/ion_heaps
-	rm -rf Release/finger_trace
-	rm -rf Release/FBInk-*.tar.xz
-	rm -rf Debug/*.a
-	rm -rf Debug/*.so*
-	rm -rf Debug/shared/*.o
-	rm -rf Debug/shared/*.opt.yaml
-	rm -rf Debug/shared/cutef8/*.o
-	rm -rf Debug/shared/cutef8/*.opt.yaml
-	rm -rf Debug/shared/libunibreak/src/*.o
-	rm -rf Debug/shared/libunibreak/src/*.opt.yaml
-	rm -rf Debug/shared/qimagescale/*.o
-	rm -rf Debug/shared/qimagescale/*.opt.yaml
-	rm -rf Debug/shared/utf8
-	rm -rf Debug/static/*.o
-	rm -rf Debug/static/*.opt.yaml
-	rm -rf Debug/static/cutef8/*.o
-	rm -rf Debug/static/cutef8/*.opt.yaml
-	rm -rf Debug/static/libunibreak/src/*.o
-	rm -rf Debug/static/libunibreak/src/*.opt.yaml
-	rm -rf Debug/static/qimagescale/*.o
-	rm -rf Debug/static/qimagescale/*.opt.yaml
-	rm -rf Debug/static/utf8
-	rm -rf Debug/*.o
-	rm -rf Debug/*.opt.yaml
-	rm -rf Debug/fbink
-	rm -rf Debug/button_scan
-	rm -rf Debug/rota
-	rm -rf Debug/fbdepth
-	rm -rf Debug/alt_buffer
-	rm -rf Debug/doom
-	rm -rf Debug/dump
-	rm -rf Debug/Kobo-DevCap-Test.tar.gz
-	rm -rf Debug/kx122_i2c
-	rm -rf Debug/ion_heaps
-	rm -rf Debug/finger_trace
+	rm -rf $(OUT_DIR)/*.a
+	rm -rf $(OUT_DIR)/*.so*
+	rm -rf $(OUT_DIR)/shared/*.o
+	rm -rf $(OUT_DIR)/shared/*.opt.yaml
+	rm -rf $(OUT_DIR)/shared/cutef8/*.o
+	rm -rf $(OUT_DIR)/shared/cutef8/*.opt.yaml
+	rm -rf $(OUT_DIR)/shared/libunibreak/src/*.o
+	rm -rf $(OUT_DIR)/shared/libunibreak/src/*.opt.yaml
+	rm -rf $(OUT_DIR)/shared/qimagescale/*.o
+	rm -rf $(OUT_DIR)/shared/qimagescale/*.opt.yaml
+	rm -rf $(OUT_DIR)/shared/utf8
+	rm -rf $(OUT_DIR)/static/*.o
+	rm -rf $(OUT_DIR)/static/*.opt.yaml
+	rm -rf $(OUT_DIR)/static/cutef8/*.o
+	rm -rf $(OUT_DIR)/static/cutef8/*.opt.yaml
+	rm -rf $(OUT_DIR)/static/libunibreak/src/*.o
+	rm -rf $(OUT_DIR)/static/libunibreak/src/*.opt.yaml
+	rm -rf $(OUT_DIR)/static/qimagescale/*.o
+	rm -rf $(OUT_DIR)/static/qimagescale/*.opt.yaml
+	rm -rf $(OUT_DIR)/static/utf8
+	rm -rf $(OUT_DIR)/*.o
+	rm -rf $(OUT_DIR)/*.opt.yaml
+	rm -rf $(OUT_DIR)/fbink
+	rm -rf $(OUT_DIR)/button_scan
+	rm -rf $(OUT_DIR)/rota
+	rm -rf $(OUT_DIR)/fbdepth
+	rm -rf $(OUT_DIR)/alt_buffer
+	rm -rf $(OUT_DIR)/doom
+	rm -rf $(OUT_DIR)/dump
+	rm -rf $(OUT_DIR)/Kobo-DevCap-Test.tar.gz
+	rm -rf $(OUT_DIR)/kx122_i2c
+	rm -rf $(OUT_DIR)/ion_heaps
+	rm -rf $(OUT_DIR)/finger_trace
 
 distclean: clean libunibreakclean libi2cclean libevdevclean
-	rm -rf libunibreak-staged
-	rm -rf libunibreak.built
-	rm -rf libi2c-staged
-	rm -rf libi2c.built
-	rm -rf libevdev-staged
-	rm -rf libevdev.built
+	rm -rf $(OUT_DIR)/libunibreak-staged
+	rm -rf $(OUT_DIR)/libunibreak.built
+	rm -rf $(OUT_DIR)/libi2c-staged
+	rm -rf $(OUT_DIR)/libi2c.built
+	rm -rf $(OUT_DIR)/libevdev-build
+	rm -rf $(OUT_DIR)/libevdev-staged
+	rm -rf $(OUT_DIR)/libevdev.built
 	-[[ -d .git ]] && rm -rf VERSION
 
 dist: distclean
