diff --git a/glib/gspawn.c b/glib/gspawn.c
index 69d3fec10..c598e02d0 100644
--- a/glib/gspawn.c
+++ b/glib/gspawn.c
@@ -50,6 +50,13 @@
 #ifdef __linux__
 #include <sys/syscall.h>  /* for syscall and SYS_getdents64 */
 #endif
+#ifdef __ANDROID__
+#include <android/api-level.h>
+#endif
+#if defined(__ANDROID__) && __ANDROID_API__ < 21 && !defined(SYS_getdents64)
+#include <sys/linux-syscalls.h>
+#define SYS_getdents64  __NR_getdents64
+#endif
 
 #include "gspawn.h"
 #include "gspawn-private.h"
