--- i/meson.build
+++ w/meson.build
@@ -416,7 +416,6 @@
   'memalign',
   'mmap',
   'newlocale',
-  'pipe2',
   'poll',
   'prlimit',
   'readlink',
@@ -596,14 +595,6 @@
 endif
 message('Checking whether to use statfs or statvfs .. ' + stat_func_to_use)
 
-if host_system == 'linux'
-  if cc.has_function('mkostemp',
-                     prefix: '''#define _GNU_SOURCE
-                                #include <stdlib.h>''')
-    glib_conf.set('HAVE_MKOSTEMP', 1)
-  endif
-endif
-
 osx_ldflags = []
 glib_have_os_x_9_or_later = false
 glib_have_carbon = false
@@ -658,16 +649,6 @@
   glib_conf.set('HAVE_FUTEX', 1)
 endif
 
-# Check for eventfd(2)
-if cc.links('''#include <sys/eventfd.h>
-               #include <unistd.h>
-               int main (int argc, char ** argv) {
-                 eventfd (0, EFD_CLOEXEC);
-                 return 0;
-               }''', name : 'eventfd(2) system call')
-  glib_conf.set('HAVE_EVENTFD', 1)
-endif
-
 clock_gettime_test_code = '''
   #include <time.h>
   struct timespec t;

