--- i/meson.build
+++ w/meson.build
@@ -709,7 +709,6 @@ functions = [
   'memalign',
   'mmap',
   'newlocale',
-  'pipe2',
   'poll',
   'prlimit',
   'readlink',
@@ -945,14 +944,6 @@ else
 endif
 message('Checking whether to use statfs or statvfs .. ' + stat_func_to_use)
 
-if host_system == 'linux'
-  if cc.has_function('mkostemp',
-                     prefix: '''#define _GNU_SOURCE
-                                #include <stdlib.h>''')
-    glib_conf.set('HAVE_MKOSTEMP', 1)
-  endif
-endif
-
 osx_ldflags = []
 glib_have_os_x_9_or_later = false
 glib_have_carbon = false
@@ -1014,16 +1005,6 @@ if cc.compiles('''#include <linux/futex.h>
   glib_conf.set('HAVE_FUTEX_TIME64', 1)
 endif
 
-# Check for eventfd(2)
-if cc.links('''#include <sys/eventfd.h>
-               #include <unistd.h>
-               int main (int argc, char ** argv) {
-                 eventfd (0, EFD_CLOEXEC);
-                 return 0;
-               }''', name : 'eventfd(2) system call')
-  glib_conf.set('HAVE_EVENTFD', 1)
-endif
-
 # Check for ppoll(2)
 if cc.links('''#define _GNU_SOURCE
                #include <poll.h>
@@ -2113,9 +2094,6 @@ else
   if cc.has_header_symbol('pthread.h', 'pthread_getname_np', prefix : pthread_prefix)
     glib_conf.set('HAVE_PTHREAD_GETNAME_NP', 1)
   endif
-  if cc.has_header_symbol('pthread.h', 'pthread_getaffinity_np', prefix : pthread_prefix)
-    glib_conf.set('HAVE_PTHREAD_GETAFFINITY_NP', 1)
-  endif
 
   # Assume that pthread_setname_np is available in some form; same as configure
   if cc.links(pthread_prefix + '''
