--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -116,17 +116,12 @@
 
 find_package(ZLIB REQUIRED)
 
-if(UNIX
-   AND NOT (APPLE OR BEOS OR HAIKU)
-   AND NOT EMSCRIPTEN)
-  find_library(M_LIBRARY m)
-  if(M_LIBRARY)
+include(CheckLibraryExists)
+check_library_exists(m floor "" M_LIBRARY)
+if(M_LIBRARY)
     set(M_LIBRARY m)
-  else()
-    set(M_LIBRARY "")
-  endif()
 else()
-  # libm is not available or not needed.
+    set(M_LIBRARY "")
 endif()
 
 if(PNG_HARDWARE_OPTIMIZATIONS)
